src/object-pool.ts

```diff
@@ -1,10 +1,287 @@
+	public state: ArmorObjectPoolState<T>;
+	public startSize: number = 10;
+	public readonly poolObj: ArmorObjectPoolInstance<T>;
 
-	constructor(options?: ArmorObjectPoolOptions) {
-		this.state = new ArmorObjectPoolState(options);
+	constructor(poolObj: ArmorObjectPoolInstance<T>, options?: ArmorObjectPoolOptions<T>) {
+		if (typeof poolObj !== 'function') {
+			throw new Error('Must have a comparator function for priority queue to operate properly');
+		}
+		this.poolObj = poolObj;
+
+		this.state = {
+			type: 'opState',
+			elements: [],
+			objectCount: 0,
+			maxSize: 1000,
+			autoIncrease: false,
+			increaseFactor: 2,
+			increaseBreakPoint: 0.8
+		};
+
+		if (options) {
+			this.parseOptions(options);
+		}
+
+		this.increase(this.startSize);
+	}
+	public parseOptions(options: ArmorObjectPoolOptions<T>): void {
+		if (!options) {
+			return;
+		}
+
+		if (options.state !== undefined) {
+			this.parseOptionsState(options.state);
+		}
+
+		if (options.startSize !== undefined) {
+			this.parseOptionsStartSize(options.startSize);
+		}
+	}
+	public parseOptionsStartSize(startSize: number): void {
+		if (!this.isInteger(startSize)) {
+			return;
+		}
+		if (startSize < 0) {
+			return;
+		}
+
+		this.startSize = startSize;
+	}
+	public parseOptionsState(state: ArmorObjectPoolState<T> | string): void {
+		if (!state) {
+			return;
+		}
+
+		let result: ArmorObjectPoolState<T> | null = null;
+
+		if (typeof state === 'string') {
+			result = this.parse(state)!;
+
+			if (!result) {
+				throw new Error('state is not valid');
+			}
+		} else {
+			result = state;
+
+			if (state.type !== 'opState') {
+				throw new Error('state must be an ArmorObjectPoolState');
+			}
+			if (typeof result.autoIncrease !== 'boolean') {
+				throw new Error('state autoIncrease must be a boolean');
+			}
+
+			if (!this.isInteger(result.objectCount)) {
+				throw new Error('state objectCount must be an integer');
+			}
+			if (!this.isInteger(result.maxSize)) {
+				throw new Error('state maxSize must be an integer');
+			}
+
+			if (typeof result.increaseFactor !== 'number' || result.increaseFactor < 0) {
+				throw new Error('state increaseFactor must be a positive number');
+			}
+
+			const between0and1 = state.increaseBreakPoint >= 0 && state.increaseBreakPoint <= 1;
+			if (typeof result.increaseBreakPoint !== 'number' || !between0and1) {
+				throw new Error('state increaseBreakPoint must be a number between 0 and 1');
+			}
+		}
+
+		this.state.autoIncrease = result.autoIncrease;
+		this.state.maxSize = result.maxSize;
+		this.state.increaseFactor = result.increaseFactor;
+		this.state.increaseBreakPoint = result.increaseBreakPoint;
+
+		this.startSize = result.objectCount;
+	}
+
+	public utilization(n: number = 0): number {
+		if (!this.isValidState(this.state)) {
+			return NaN;
+		}
+		if (this.state.objectCount === 0) {
+			return Infinity;
+		}
+
+		let num: number = n;
+		if (isNaN(num)) {
```

Comment:
Good check for `isNaN()` since it's a valid number. Unfortunately TS does not guarantee the types if casted (e.g. `'string' as number`) or if called from JS. Add an additional number typecheck here to guard against that as it will throw, or worse cause undefined behavior.
```isNaN(num) || typeof num !== 'number'```